lib/compat/wordpress-6.0/block-patterns.php

```diff
@@ -39,3 +39,92 @@ function gutenberg_register_remote_theme_patterns() {
 }
 
 add_action( 'init', 'gutenberg_register_remote_theme_patterns' );
+
+/**
+ * Register any patterns that the active theme may provide under its
+ * `./patterns/` directory. Each pattern is defined as a PHP file and defines
+ * its metadata using plugin-style headers. The minimum required definition is:
+ *
+ *     /**
+ *      * Pattern Name: My Pattern
+ *      *
+ *
+ * The output of the PHP source corresponds to the content of the pattern, e.g.:
+ *
+ *     <main><p><?php echo "Hello"; ?></p></main>
+ *
+ * If applicable, this will collect from both parent and child theme.
+ *
+ * @since 6.0.0
+ * @access private
+ * @internal
+ */
+function gutenberg_register_theme_block_patterns() {
+	$default_headers = array(
+		'title'         => 'Pattern Name',
+		'description'   => 'Description',
+		'viewportWidth' => 'Viewport Width',
+		'categories'    => 'Categories',
+		'keywords'      => 'Keywords',
+		'blockTypes'    => 'Block Types',
+	);
+
+	// Register patterns for the active theme, for both parent and child theme,
+	// if applicable.
+	foreach ( wp_get_active_and_valid_themes() as $theme ) {
+		$dirpath = $theme . '/patterns/';
+		if ( file_exists( $dirpath ) ) {
+			$files = glob( $dirpath . '*.php' );
+			if ( $files ) {
+				foreach ( $files as $file ) {
+					// Parse pattern slug from file name.
+					if ( ! preg_match( '#/(?P<slug>[A-z0-9_-]+)\.php$#', $file, $matches ) ) {
+						continue; // FIXME: Consider logging notice.
+					}
+					// Example name: twentytwentytwo/query-grid-posts.
+					$pattern_name = get_stylesheet() . '/' . $matches['slug'];
```

Comment:
Nice touch with the `wp_get_active_and_valid_themes` so we consider both the parent and the child patterns. This is great. What do you think about the following use case?

- the parent theme called `my-theme` has a pattern called `footer-blog.php`
- the child theme called `my-theme-child` has a pattern called `footer-blog.php`
- as it works now, we register two patterns: `my-theme/footer-blog`, `my-theme-child/footer-blog`

I'm not completely sure about the expectations here: should a child theme be allowed to overwrite a pattern from the parent theme?

By inspectiing how the `register_block_pattern` API works I learned that if you register a pattern with the same name twice the later wins. There's also the assumption that child themes being a way to "override" things from the parent. Taking those thing into account I'm inclined to think that we should allow child themes to override the parent patterns. Thoughts?

Reply:
it makes sense to me that the child theme would override the parent's rather than duplicate them

Reply:
Same opinion from me.

Reply:
It looks like you will have to split the process into two steps. First, that collects the files and removes those overridden. Next, you would register all unique patterns.

Reply:
Give cb99311e6c432defc2b5f9689c988bb02996952b a go.

Reply:
@gziolo: do you foresee issues with the above? In my mind, it makes sense. If need be, we could clarify in `register_block_pattern` that re-registering a pattern under the same key replaces the existing definition.

Reply:
Can you use a check if a pattern was already registered?

Reply:
Yes, there is an API for that:
https://github.com/WordPress/wordpress-develop/blob/26bb0d7479e1dd08dd982513ee6bee94f4c6f13d/src/wp-includes/block-patterns.php#L124

We probably could add a simple function wrapper to make it simpler to use.

Reply:
-> 776c35d490